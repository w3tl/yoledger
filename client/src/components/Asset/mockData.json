[{
  "id": "1",
  "name": "Bank Card",
  "balance": 0,
  "type": "ASSET"
}, {
  "id": "2",
  "name": "Cash",
  "balance": 10,
  "type": "ASSET"
}, {
  "id": "3",
  "name": "Bank Card 2",
  "balance": 150,
  "type": "ASSET"
}]
