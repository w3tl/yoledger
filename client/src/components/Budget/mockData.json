{
  "__typename": "BudgetPlan",
  "accounts": [{
    "__typename": "Account",
    "id": "1",
    "name": "Food"
  }, {
    "__typename": "Account",
    "id": "2",
    "name": "Train"
  }],
  "budgets": [
    {
      "id": "1",
      "account": {
        "id": "2",
        "name": "Train"
      },
      "date": "2018-06-10",
      "amount": 11,
      "balance": 0
    }, {
      "id": "2",
      "account": {
        "id": "1",
        "name": "Food"
      },
      "date": "2018-06-10",
      "amount": 12,
      "balance": 0
    }, {
      "id": "3",
      "account": {
        "id": "2",
        "name": "Train"
      },
      "date": "2018-06-25",
      "amount": 13,
      "balance": 0
    }, {
      "id": "4",
      "account": {
        "id": "1",
        "name": "Food"
      },
      "date": "2018-07-10",
      "amount": 14,
      "balance": 0
    }, {
      "id": "5",
      "account": {
        "id": "1",
        "name": "Food"
      },
      "date": "2018-07-25",
      "amount": 16,
      "balance": 0
    }, {
      "id": "6",
      "account": {
        "id": "2",
        "name": "Train"
      },
      "date": "2018-07-25",
      "amount": 17,
      "balance": 0
    }, {
      "id": "7",
      "account": {
        "id": "2",
        "name": "Train"
      },
      "date": "2018-08-10",
      "amount": 19,
      "balance": 0
    }, {
      "id": "8",
      "account": {
        "id": "1",
        "name": "Food"
      },
      "date": "2018-08-10",
      "amount": 20,
      "balance": 0
    }, {
      "id": "9",
      "account": {
        "id": "2",
        "name": "Train"
      },
      "date": "2018-08-25",
      "amount": 21,
      "balance": 0
    }
  ]
}
