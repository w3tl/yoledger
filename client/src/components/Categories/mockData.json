{
  "expenses": [{
    "id": "21",
    "name": "Food",
    "type": "EXPENSE"
  }, {
    "id": "22",
    "name": "Beer",
    "type": "EXPENSE"
  }, {
    "id": "23",
    "name": "Train",
    "type": "EXPENSE"
  }],
  "incomes": [{
    "id": "31",
    "name": "Salary",
    "type": "INCOME"
  }, {
    "id": "32",
    "name": "Debts",
    "type": "INCOME"
  }, {
    "id": "33",
    "name": "Cashback",
    "type": "INCOME"
  }]
}
