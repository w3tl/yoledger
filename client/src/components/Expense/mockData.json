[{
  "id": "21",
  "name": "Food"
}, {
  "id": "22",
  "name": "Beer"
}, {
  "id": "23",
  "name": "Train"
}]
