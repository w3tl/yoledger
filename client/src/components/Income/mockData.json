[{
  "id": "31",
  "name": "Salary"
}, {
  "id": "32",
  "name": "Debts"
}, {
  "id": "33",
  "name": "Dividends"
}]
