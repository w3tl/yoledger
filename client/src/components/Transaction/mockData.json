[{
  "__typename": "Transaction",
  "id": "1",
  "amount": 10,
  "source": {
    "__typename": "Account",
    "id": "2",
    "name": "Cash"
  },
  "destination": {
    "__typename": "Account",
    "id": "21",
    "name": "Food"
  },
  "date": "2018-01-01T00:00:00.000Z"
}, {
  "__typename": "Transaction",
  "id": "2",
  "amount": 10,
  "source": {
    "__typename": "Account",
    "id": "2",
    "name": "Cash"
  },
  "destination": {
    "__typename": "Account",
    "id": "1",
    "name": "Bank Card"
  },
  "date": "2018-04-01T00:00:00.000Z"
}, {
  "__typename": "Transaction",
  "id": "3",
  "amount": 10,
  "source": {
    "__typename": "Account",
    "id": "1",
    "name": "Bank Card"
  },
  "destination": {
    "__typename": "Account",
    "id": "21",
    "name": "Food"
  },
  "date": "2018-01-06T00:00:00.000Z"
}]
