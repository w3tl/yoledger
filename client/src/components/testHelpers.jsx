import React from 'react';
import { shallow, mount } from 'enzyme';
import { MockedProvider } from 'react-apollo/test-utils';

export const wait = (ms = 0) => new Promise(resolve => setTimeout(resolve, ms));

export const withProvider = Component => props => (
  <MockedProvider {...props}>
    <Component />
  </MockedProvider>
);

export const testRenderWithoutError = (Component) => {
  it('renders without error', async () => {
    const wrapper = shallow(Component);
    await wait();
    wrapper.update();
    expect(wrapper).toMatchSnapshot();
  });
};

export const testLoadingState = (Component) => {
  it('should render loading state initially', () => {
    const wrapper = mount(Component);
    expect(wrapper.text()).toEqual('Loading...');
  });
};

export const testErrorUI = (Component) => {
  it('should show error UI', async () => {
    const wrapper = mount(Component);
    await wait(0); // wait for response
    expect(wrapper.update().text()).toContain('Error');
  });
};

// THNX: http://blog.blakesimpson.co.uk/read/98-how-to-stub-date-in-jest-tests-building-a-stubdate-helper
export const stubDate = (fixedDate) => {
  let originalDate;

  beforeAll(() => {
    originalDate = Date;

    global.Date = class extends Date {
      constructor() {
        super();

        return fixedDate;
      }
    };
  });

  afterAll(() => {
    global.Date = originalDate;
  });
};
