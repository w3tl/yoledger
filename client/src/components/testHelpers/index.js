import React from 'react';
import { shallow, mount } from 'enzyme';
import { ApolloProvider } from 'react-apollo';
import { ApolloClient } from 'apollo-client';
import { InMemoryCache } from 'apollo-cache-inmemory';
import { SchemaLink } from 'apollo-link-schema';
import { makeExecutableSchema } from 'graphql-tools';
import typeDefs from './typeDefs';
import resolvers from './resolvers';

export const wait = (ms = 0) => new Promise(resolve => setTimeout(resolve, ms));

const executableSchema = makeExecutableSchema({
  typeDefs,
  resolvers,
});

const cache = new InMemoryCache();
const link = new SchemaLink({ schema: executableSchema });
export const client = new ApolloClient({
  link,
  cache,
});

export const withProvider = Component => props => (
  <ApolloProvider client={client}>
    <Component client={client} />
  </ApolloProvider>
);

export const testRenderWithoutError = (Component) => {
  it('renders without error', async () => {
    const wrapper = shallow(Component);
    await wait();
    wrapper.update();
    expect(wrapper).toMatchSnapshot();
  });
};

export const testLoadingState = (Component) => {
  it('should render loading state initially', () => {
    const wrapper = mount(Component);
    expect(wrapper.text()).toEqual('Loading...');
  });
};

export const testErrorUI = (Component) => {
  it('should show error UI', async () => {
    const wrapper = mount(Component);
    await wait(0); // wait for response
    expect(wrapper.update().text()).toContain('Error');
  });
};

// THNX: http://blog.blakesimpson.co.uk/read/98-how-to-stub-date-in-jest-tests-building-a-stubdate-helper
export const stubDate = (fixedDate) => {
  const date = new Date(fixedDate);
  let originalDate;

  beforeAll(() => {
    originalDate = Date;

    global.Date = class extends Date {
      constructor() {
        super();

        return date;
      }
    };
  });

  afterAll(() => {
    global.Date = originalDate;
  });
};
