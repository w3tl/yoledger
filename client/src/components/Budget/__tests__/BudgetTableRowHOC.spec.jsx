/* eslint-disable react/prop-types */
import React from 'react';
import { mount } from 'enzyme';
import RawBudgetTableRow, { updateAfterUpsert } from '../BudgetTableRowHOC';
import { accounts, budgets } from '../mockData';
import { wait, withProvider, testLoadingState } from '../../testHelpers';

const periods = [new Date('2018-06-10'), new Date('2018-06-25'), new Date('2018-07-10')];

jest.mock('../BudgetTableCell', () => ({ onUpdate, amount }) => (
  <td>
    <input name="amount" value={amount} onChange={() => {}} />
    <button name="updateAmount" type="button" onClick={() => onUpdate({ amount: 43 })} />
  </td>
));

const variables = {
  account: accounts[0],
  dateStart: periods[0].toISOString(),
  dateEnd: periods[periods.length - 1].toISOString(),
};

const BudgetTableRow = withProvider(props => (
  <table>
    <tbody>
      <RawBudgetTableRow periods={periods} {...variables} {...props} />
    </tbody>
  </table>
));

describe('BudgetTableRowHOC', () => {
  testLoadingState(<BudgetTableRow />);

  describe('updateAfterUpsert', () => {
    it('should append new budget', () => {
      const writeQuery = jest.fn();
      const cache = {
        readQuery: jest.fn().mockReturnValue({ budget: [budgets[0]] }),
        writeQuery,
      };
      const data = { upsertBudget: { budget: { ...budgets[0], id: 'new' } } };
      updateAfterUpsert(variables)(cache, { data });
      expect(cache.readQuery.mock.calls[0][0].variables).toMatchSnapshot('readQuery');
      expect(writeQuery.mock.calls[0][0].data).toMatchSnapshot('writeQuery');
    });

    it('should write existing budget array if budget found', () => {
      const writeQuery = jest.fn();
      const cache = {
        readQuery: jest.fn().mockReturnValue({ budget: [budgets[0]] }),
        writeQuery,
      };
      const data = { upsertBudget: { budget: budgets[0] } };
      updateAfterUpsert(variables)(cache, { data });
      expect(cache.readQuery.mock.calls[0][0].vaariables).toMatchSnapshot('readQuery');
      expect(writeQuery.mock.calls[0][0].data).toMatchSnapshot('writeQuery');
    });
  });

  it('should have props', async () => {
    const wrapper = mount(<BudgetTableRow />);
    await wait(1000);
    wrapper.update();
    expect(wrapper.find('BudgetTableRow').prop('account')).toBeDefined();
    expect(wrapper.find('BudgetTableRow').prop('budget').length).toBeGreaterThan(0);
  });
});
