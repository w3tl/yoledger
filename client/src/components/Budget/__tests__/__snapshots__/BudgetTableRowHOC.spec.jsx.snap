// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BudgetTableRowHOC updateAfterUpsert should append new budget: readQuery 1`] = `
Object {
  "account": "Food",
  "dateEnd": "2018-07-10T00:00:00.000Z",
  "dateStart": "2018-06-10T00:00:00.000Z",
}
`;

exports[`BudgetTableRowHOC updateAfterUpsert should append new budget: writeQuery 1`] = `
Object {
  "budget": Array [
    Object {
      "account": Object {
        "id": "2",
        "name": "Train",
      },
      "amount": 11,
      "balance": 0,
      "date": "2018-06-10",
      "id": "1",
    },
    Object {
      "account": Object {
        "id": "2",
        "name": "Train",
      },
      "amount": 11,
      "balance": 0,
      "date": "2018-06-10",
      "id": "new",
    },
  ],
}
`;

exports[`BudgetTableRowHOC updateAfterUpsert should write existing budget array if budget found: readQuery 1`] = `undefined`;

exports[`BudgetTableRowHOC updateAfterUpsert should write existing budget array if budget found: writeQuery 1`] = `
Object {
  "budget": Array [
    Object {
      "account": Object {
        "id": "2",
        "name": "Train",
      },
      "amount": 11,
      "balance": 0,
      "date": "2018-06-10",
      "id": "1",
    },
  ],
}
`;
